numpy/lib/function_base.py

```diff
@@ -3929,21 +3929,40 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        r_above = x1 + (x2 - x1) * weights_above
+        r_below = x2 - (x2 - x1) * weights_below
 
         # ensure axis with q-th is first
-        x1 = np.moveaxis(x1, axis, 0)
-        x2 = np.moveaxis(x2, axis, 0)
+        r_above = np.moveaxis(r_above, axis, 0)
+        r_below = np.moveaxis(r_below, axis, 0)
 
         if zerod:
-            x1 = x1.squeeze(0)
-            x2 = x2.squeeze(0)
+            weights_above = weights_above.squeeze(0)
+            weights_below = weights_below.squeeze(0)
+            r_above = r_above.squeeze(0)
+            r_below = r_below.squeeze(0)
+
+        r = np.where(weights_above < 0.5, r_above, r_below)
 
         if out is not None:
-            r = add(x1, x2, out=out)
+            if out.ndim == 0:
```

Comment:
I should admit that the following block is really messy. I lack knowledge regarding 0-d array and how to make in-place computation. I would appreciate any comment on how to improve this.